### DISCUSSION TITLE
Any viable Assembly-level optimizations ideas?

### DISCUSSION THREAD

**Discussion by KaelaSavia** [ggml-org/llama.cpp#6981] [Ideas]
To go further beyond potential performance, I've been wondering if there's something we could do at assembly level. Like placing hot-functions/frequently called functions next to eachother in .code segment, perhaps that could improve performance by lessening distance the cpu has to jump to assembly instructions to execute

Other than that, I do wonder whether performance would be better on x86 versus x86-64 instruction set. Some of my projects had significant improvements on just x86, so perhaps having parts of architecture in x86 binary while others in x86-64 binary could improve performance due to lessening pointer arithmetics and such

It's just random brain thoughts im having but I think it's fun direction to explore

**Comment by ivanstepanovftw** [comment]
On a CPU, the bottleneck is not the number of instructions per second, but the bandwidth of the RAM.
